src/fsharp/ErrorLogger.fs

```diff
@@ -686,3 +686,103 @@ let internal languageFeatureNotSupportedInLibraryError (langVersion: LanguageVer
     let suggestedVersionStr = langVersion.GetFeatureVersionString langFeature
     error (Error(FSComp.SR.chkFeatureNotSupportedInLibrary(featureStr, suggestedVersionStr), m))
+
+/// This represents the thread-local state established as each task function runs as part of the build.
+///
+/// Use to reset error and warning handlers.
+type CompilationGlobalsScope(errorLogger: ErrorLogger, phase: BuildPhase) = 
+    let unwindEL = PushErrorLoggerPhaseUntilUnwind(fun _ -> errorLogger)
+    let unwindBP = PushThreadBuildPhaseUntilUnwind phase
+
+    member _.ErrorLogger = errorLogger
+    member _.Phase = phase
+
+    // Return the disposable object that cleans up
+    interface IDisposable with
+        member d.Dispose() =
+            unwindBP.Dispose()         
+            unwindEL.Dispose()
+
+[<Sealed>]
+type AsyncErrorLoggerBuilder() =
+
+    member _.Zero () : Async<unit> = 
+        async { 
+            ()
+        }
+
+    member _.Delay (f: unit -> Async<'T>) = f()   
+
+    member _.Return value =
+        async { 
+            return value
+        }
+
+    member _.ReturnFrom (computation:Async<_>) = computation
+
+    member _.Bind (computation: Async<'a>, binder: 'a -> Async<'b>) : Async<'b> =
+        async {
+            let errorLogger = CompileThreadStatic.ErrorLogger
+            let phase = CompileThreadStatic.BuildPhase
+            try
+                let! res = computation
+                CompileThreadStatic.ErrorLogger <- errorLogger
+                CompileThreadStatic.BuildPhase <- phase
+                return! binder res
+            finally
+                CompileThreadStatic.ErrorLogger <- errorLogger
+                CompileThreadStatic.BuildPhase <- phase
+        }
+
+    member _.TryWith(computation: Async<'T>, binder: exn -> Async<'T>) : Async<'T> =
+        async {
+            let errorLogger = CompileThreadStatic.ErrorLogger
+            let phase = CompileThreadStatic.BuildPhase
+            try
+                try
+                    return! computation
+                with
+                | ex ->  
+                    CompileThreadStatic.ErrorLogger <- errorLogger
+                    CompileThreadStatic.BuildPhase <- phase
+                    return! binder ex
+            finally
+                CompileThreadStatic.ErrorLogger <- errorLogger
+                CompileThreadStatic.BuildPhase <- phase
+        }
+
+    member _.Using(value: CompilationGlobalsScope, binder: CompilationGlobalsScope -> Async<'U>) =
+        async {
+            CompileThreadStatic.ErrorLogger <- value.ErrorLogger
+            CompileThreadStatic.BuildPhase <- value.Phase
+
+            try
+                return!
+                    try
+                        binder value 
+                    with 
+                    | _ -> 
+                        (value :> IDisposable).Dispose()
+                        reraise()
+            finally
+                (value :> IDisposable).Dispose()
```

Comment:
This will dispose twice when an exception is raised.  The try/finally should be enough.  

Indeed this should be enough:

```fsharp
        async {
            CompileThreadStatic.ErrorLogger <- value.ErrorLogger
            CompileThreadStatic.BuildPhase <- value.Phase

            use r = value
            return! binder value 
        }
```

However I don't understand why `Using` needs to re-establish the CompileThreadStatic at all?

Reply:
Note also that in a typical "Using" the resoure is legitimately allowed to be null